main to CLI

Anwesh1 opened the pull request:
# Overview

Have 2 required arguments set up for start and end dates respectively
```main.py --help``` shows this:
```bash
C:\Users\anwes\Desktop\pupul\BaderLab\Pubmed\semantic_git\pubmed-dl\pubmed_dl>py main.py --help
Usage: main.py [OPTIONS] START END

Arguments:
  START  [required]
  END    [required]

Options:
  --install-completion [bash|zsh|fish|powershell|pwsh]
                                  Install completion for the specified shell.
  --show-completion [bash|zsh|fish|powershell|pwsh]
                                  Show completion for the specified shell, to
                                  copy it or customize the installation.

  --help                          Show this message and exit.
```
for the testing run any dates:
```main.py 2021/02/05 2021/02/05```
should work and write to the ```test.json``` file

## Closes

Closes #12.

Anwesh1 committed:
Update main.py

Anwesh1 committed:
Update main.py

JohnGiorgi commented:
This is a good start. I have two main suggestions:

- Document `main` and its arguments (see [here](https://typer.tiangolo.com/tutorial/first-steps/#document-your-cli-app)).
- Consider adding a progress bar (see [here](https://typer.tiangolo.com/tutorial/progressbar/)). This ties into what we talked about with benchmarking and will make that much easier. If this is tricky feel free to break it off into its own PR.

Some general SE advice: Try making your commit messages informative. Lots of great tutorials on this ([example](https://www.freecodecamp.org/news/writing-good-commit-messages-a-practical-guide/)).

JohnGiorgi commented on pubmed_dl/main.py:
```diff
@@ -1,15 +1,16 @@
-    start = input("Enter start date (YYYY/MM/DD): ")
-    end = input("Enter end date (YYYY/MM/DD): ")
-    val = get_list_pmid(start,end)
+def main(start: str, end: str):
```
Could we think of more informative names for these variables? E.g. `start_date` and `end_date`. Think of variable names as your first layer of documentation.

JohnGiorgi commented on pubmed_dl/main.py:
```diff
@@ -1,15 +1,16 @@
+def main(start: str, end: str):
+    start_date = start
+    end_date = end
+    val = get_list_pmid(start_date,end_date)
```
Ditto. `val` is not a good variable name. `pmids` would be better, for example.

jvwong commented:
Just FYI: the `main` branch has changed because of #18 so you should:

1. Update the main branch: `git checkout main` and then `git pull` to update it
2. Update your branch: `git checkout cli_main` then `git merge main`

If there are conflicts, you'll have to resolve them which is always a pain. If not, then you're good.

Anwesh1 committed:
Update main.py

JohnGiorgi commented on pubmed_dl/main.py:
```diff
@@ -1,15 +1,16 @@
+    start_date = start
+    end_date = end
+    val = get_list_pmid(start_date,end_date)
     with open("test.json", "w") as f:
```
`"test.json"` is hardcoded. It should be one of the arguments to `main` that a user can specify at the command line. Remember that if they specify a directory you will have to create it. I use the following code typically:

```python
from pathlib import Path

# output_filepath provided as command line argument as string...

output_filepath: Path = Path(output_filepath)
output_filepath.parents[0].mkdir(parents=True, exist_ok=True)
```

Anwesh1 committed:
Typer 3 arguments with filename optional

Need to add a progress bar feature and the ability to create a directory if requested as an argument.

Anwesh1 committed:
Merge branch 'main' into cli_main

Anwesh1 committed:
Update main.py

JohnGiorgi commented on pubmed_dl/main.py:
```diff
@@ -1,38 +1,55 @@
+logging.basicConfig(level=settings.loglevel)
+
+
+def main(start: str, end: str, request_file: Optional[str] = typer.Argument(None)):
```
I wouldn't make the output file path a default unless there's a good reason for it. Output file paths are almost always required arguments for CLIs.

Not providing it usually means it will be printed to `stdout`, but we don't want to do that here.

Anwesh1 replied:
I made it optional in the sense that if it's not provided then it writes to ```test.json```. However, if it is provided, then it goes there. As of right now it always writes to a file, it's either to ```test.json```(if nothing is entered) or it is what they request. But I can change it to required instead of optional, it's your call.

I'm trying to figure out how the directory thing works now, the one you sent above.

JohnGiorgi replied:
I would make it required unless there's a good argument for `"test.json"` being the default. 

The `Path` code above just creates the directory if it doesn't exist. If a user provides the path `some/dir/output.json` and `some/dir` doesn't exist there will be an error thrown. The above code solves that by creating `some/dir` if it doesn't already exist.

Anwesh1 replied:
okay let me try that. Also I'll make the argument required

Anwesh1 replied:
Done

Anwesh1 committed:
Update main.py

Anwesh1 commented:
So far we got some good updates. Only the progress bar feature left to add. Otherwise, we can merge this and make the progress bar another PR. What would you guys prefer?

jvwong commented:
Error on typing  at command line `pubmed-dl`:

`  File "....pubmed-dl/pubmed_dl/main.py", line 9, in <module>
    import typer
ModuleNotFoundError: No module named 'typer'
`

Also is there a way to print out a useage message, i.e. how to use ?

JohnGiorgi commented:
> Error on typing at command line `pubmed-dl`:
> 
> ` File "....pubmed-dl/pubmed_dl/main.py", line 9, in <module> import typer ModuleNotFoundError: No module named 'typer'`
> 
> Also is there a way to print out a useage message, i.e. how to use ?

@Anwesh1 Will you add `typer` to `setup.py` [here](https://github.com/PathwayCommons/pubmed-dl/blob/ade1bfaab567eff09c331548fab848d9788fd44f/setup.py#L31).

@jvwong Documentation should be printed if you call `pubmed-dl --help`. @Anwesh1 Maybe update the readme to show this.

JohnGiorgi commented:
> So far we got some good updates. Only the progress bar feature left to add. Otherwise, we can merge this and make the progress bar another PR. What would you guys prefer?

I would address the outstanding comments and leave the progress bar for its own PR. Open an issue to track it.

Anwesh1 commented:
Quick question, how do I check what version for typer? so I can add on to the ```setup.py```

JohnGiorgi commented:
> Quick question, how do I check what version for typer? so I can add on to the `setup.py`

`pip freeze | grep "typer"`. Use the `>=` notation of the [existing dependencies](https://github.com/PathwayCommons/pubmed-dl/blob/ade1bfaab567eff09c331548fab848d9788fd44f/setup.py#L31).

Anwesh1 committed:
Update README.md

Anwesh1 committed:
Update setup.py

Anwesh1 commented:
Should be good, let me know if I can merge.

JohnGiorgi commented on pubmed_dl/main.py:
```diff
@@ -1,38 +1,53 @@
+    start_time = time.time()
+    output_filepath: Path = Path(request_file)
+    output_filepath.parents[0].mkdir(parents=True, exist_ok=True)
+    filename = request_file
```
Why create a third variable? Just use `output_filepath` henceforth (e.g. in the call to `with open`)

JohnGiorgi commented on README.md:
````diff
@@ -62,23 +62,54 @@ An example usage is by getting the list of PMID's from ```get_list_pmid``` and t
 
 ```bash
-pubmed-dl
+main.py --help
````
Not sure why we changed this back to `main.py`? `pubmed-dl` is better because once a user installs our package they can call it from anywhere. Calling just `main.py` assumes they have cloned this locally and are in the correct folder.

Anwesh1 replied:
I'm not able to run it as pubmed-dl, is that because I havent downloaded it as a package?

Anwesh1 replied:
how would the commands run then? `pubmed-dl --help` ??

JohnGiorgi replied:
Something has broke the `pubmed-dl` command in this PR. I will have to debug later unless you want to take this on.

Anwesh1 replied:
I never used the `pubmed-dl` command before so I'm not sure where to check for that issue. I would appreciate your help on this matter.

JohnGiorgi replied:
Just merge and open another issue to track this, I'll have to get to it later.

JohnGiorgi commented on README.md:
````diff
@@ -62,23 +62,54 @@ An example usage is by getting the list of PMID's from ```get_list_pmid``` and t
+This prints out:
 
-Program requires two dates to be inputted for start and end
+```bash
+Usage: main.py [OPTIONS] START END REQUEST_FILE
+
+  Main requires 3 arguments.
+
+  The first one is start date (eg: YYYY/MM/DD).
+
+  The second one is end date (eg: 2021/02/05).
+
+  The third one is the filename where the data needs to be written to.
+
+Arguments:
+  START         [required]
+  END           [required]
+  REQUEST_FILE  [required]
+
+Options:
+  --install-completion [bash|zsh|fish|powershell|pwsh]
+                                  Install completion for the specified shell.
+  --show-completion [bash|zsh|fish|powershell|pwsh]
+                                  Show completion for the specified shell, to
+                                  copy it or customize the installation.
+
+  --help                          Show this message and exit.
+```
+
+Program requires three arguments to be inputted by the user. 
+The first one is the start date in the ```YYYY/MM/DD``` format (e.g.: ```2021/02/04```). 
+The second one is the end date in the ```YYYY/MM/DD``` format (e.g.: ```2021/02/05```). 
+The third one is the filename where the data needs to be written to (e.g.: ```test.json```).
````
Not really important but I wouldn't copy-paste this here as it's subject to change. I would just say something like "call `pubmed-dl --help` for usage details".

JohnGiorgi commented on README.md:
````diff
@@ -62,23 +62,54 @@ An example usage is by getting the list of PMID's from ```get_list_pmid``` and t
+```
+
+Program requires three arguments to be inputted by the user. 
+The first one is the start date in the ```YYYY/MM/DD``` format (e.g.: ```2021/02/04```). 
````
``` are used to open multi-line code blocks. ` is used for inline code formatting.

Anwesh1 committed:
Update README.md

Anwesh1 committed:
Update README.md

Anwesh1 commented:
hey @JohnGiorgi is the updated `README.md` right? especially the ones where I put the example commands for the command line?

Anwesh1 committed:
Update main.py

JohnGiorgi commented on pubmed_dl/main.py:
```diff
@@ -1,38 +1,53 @@
+    output_filepath: Path = Path(request_file)
+    output_filepath.parents[0].mkdir(parents=True, exist_ok=True)
+    filename = request_file
+    start_date = start
+    end_date = end
```
This is an anti-pattern. No reason to create two new variables with slightly different names. Just call the parameters `start_date` and `end_date` to begin with.

Anwesh1 replied:
sure, done

Anwesh1 committed:
Update main.py

JohnGiorgi commented:
@Anwesh1 This might have got buried in the above comments but I would get in the habit of writing informative commit messages. See ([here](https://www.freecodecamp.org/news/writing-good-commit-messages-a-practical-guide/)).

Anwesh1 commented:
@JohnGiorgi sorry about that, I will change that from the next commit
